src/test/kotlin/org/rust/ide/hints/RsInlayParameterHintsProviderTest.kt

```diff
@@ -46,20 +47,95 @@ class RsInlayParameterHintsProviderTest : RsTestBase() {
         }
         fn main() {
             let s = S;
-            S::foo(s, <caret>0);
+            S::foo(s, /*caret*/0);
         }    //^
     """, "arg:", 0)
 
     fun testLetDecl() = checkByText<RsLetDecl>("""
         struct S;
         fn main() {
-            let s<caret> = S;
+            let s/*caret*/ = S;
         }  //^
     """, ": S", 0)
 
+    fun `test smart hint same parameter name`() = checkByText<RsCallExpr>("""
+        fn foo(arg: u32, arg2: u32) {}
+        fn main() {
+            let arg = 0;
+            foo(arg, /*caret*/1);
+        } //^
+    """, "arg2:", 0)
+
+    fun `test smart hint method start with set`() = checkNoHint<RsMethodCallExpr>("""
+        struct S;
+        impl S {
+            fn set_foo(self, arg: u32) {}
+        }
+        fn main() {
+            let s = S;
+            s.set_foo(1);
+        } //^
+    """)
+
+    fun `test smart hint self call start with set`() = checkNoHint<RsCallExpr>("""
+        struct S;
+        impl S {
+            fn set_foo(self, arg: u32) {}
+        }
+        fn main() {
+            let s = S;
+            S::set_foo(s, /*caret*/0);
+        } //^
+    """)
+
+    fun `test smart hint same function name and single parameter`() = checkNoHint<RsCallExpr>("""
+        fn foo(arg: u32) {}
+        fn main() {
+            let foo = 0;
+            foo(foo);
+        } //^
+    """)
+
+    fun `test smart hint same method name and single parameter`() = checkNoHint<RsMethodCallExpr>("""
+        struct S;
+        impl S {
+            fn foo(self, foo: u32) {}
+        }
+        fn main() {
+            let s = S;
+            s.foo(10);
+        }    //^
+    """)
+
+    fun `test smart hint same method name (self call) and single parameter`() = checkNoHint<RsCallExpr>("""
+        struct S;
+        impl S {
+            fn foo(self, foo: u32) {}
+        }
+        fn main() {
+            let s = S;
+            S::foo(s, 10);
+        }    //^
+    """)
+
+    fun `test lamdba type hint`() = checkByText<RsLambdaExpr>("""
+        struct S;
+        fn with_s<F: Fn(S)>(f: F) {}
+        fn main() {
+            with_s(|s/*caret*/| s.bar())
```

Comment:
We need to provide an `Fn` lang item for this to work. Take a look how `org.rust.lang.core.resolve.RsClosuresResolveTest` manages to do it without depending on the acutal standard library (it's better to avoid stdlib aware tests, because they are slower).

Bascically, we need to add `#[lang="fn"] trant Fn<Args> { type Output; }` code to this one test case.